src/dist/failure_detector/failure_detector.cpp

```diff
@@ -123,41 +126,39 @@ void failure_detector::register_master(::dsn::rpc_address target)
         dinfo("master[%s] already registered", target.to_string());
     }
 
-    send_beacon(target, now_ms());
+    send_beacon(target, now_ms(), ret.first->second.version);
 }
 
 bool failure_detector::switch_master(::dsn::rpc_address from, ::dsn::rpc_address to)
 {
+    auto it = _masters.find(from);
+    auto it2 = _masters.find(to);
+    if (it != _masters.end())
     {
-        zauto_lock l(_lock);
```

Comment:
best to add bool is_lock_held parameter so stand-alone switch_master is also possible later. otherwise, a public accessible switch_master w/o lock is not a good idea.